File: src/main/resources/csharp/runtime/java/util/Arrays.cs
namespace Couscous.java.util {
    internal static class Arrays {
        internal static java.lang.Iterable<T> asList<T>(T[] values) {
            return new _ArrayList<T>(values);
        }
    }
    
    internal class _ArrayList<T> : java.lang.Iterable<T> {
        private readonly T[] values;
        
        internal _ArrayList(T[] values) {
            this.values = values;
        }
        
        public java.util.Iterator<T> iterator() {
            return new _ArrayIterator<T>(values);
        }
    }
    
    internal class _ArrayIterator<T> : java.util.Iterator<T> {
        private readonly T[] values;
        private int nextIndex;
        
        internal _ArrayIterator(T[] values) {
            this.values = values;
            this.nextIndex = 0;
        }
        
        public bool hasNext() {
            return nextIndex < values.Length;
        }
        
        public T next() {
            return values[nextIndex++];
        }
    }
}

File: src/main/resources/org/zwobble/couscous/backends/python/runtime/java/util/Arrays.py
from ..._couscous import couscous_type, signature, Array
from .Iterator import Iterator
from .List import List


@couscous_type
class Arrays(object):
    @staticmethod
    @signature(Array, List)
    def asList(array):
        return ListAdapter(array)


@couscous_type
class ListAdapter(object):
    def __init__(self, elements):
        self._elements = elements
    
    @signature(Iterator)
    def iterator(self):
        return ListIterator(self._elements)


@couscous_type
class ListIterator(object):
    def __init__(self, elements):
        self._elements = elements
        self._next_index = 0
    
    @signature(bool)
    def hasNext(self):
        return self._next_index < len(self._elements)
    
    @signature(Iterator.T)
    def next(self):
        value = self._elements[self._next_index]
        self._next_index += 1
        return value
